draft-ietf-tls-external-psk-importer.md

```diff
@@ -79,156 +80,151 @@ when, and only when, they appear in all capitals, as shown here.
+connection, which is a tuple of (Base Key, External Identity, Hash). The associated hash
+function may be undefined.
 - Base Key: The secret value of an EPSK.
 - External Identity: The identity of an EPSK.
 - Imported Identity: The identity of a PSK as sent on the wire.
+- Target protocol: The protocol for which a PSK is imported for use.
+- Target KDF: The KDF for which a PSK is imported for use.
 
 # Key Import
 
-A key importer takes as input an EPSK with external identity 'external_identity' and base key 'epsk',
-as defined in {{terminology}}, along with an optional label, and transforms
-it into a set of PSKs and imported identities for use in a connection based on supported HashAlgorithms.
-In particular, for each supported HashAlgorithm 'hash', the importer constructs an ImportedIdentity
-structure as follows:
+A key importer takes as input an EPSK with external identity `external_identity` and base key `epsk`,
+as defined in {{terminology}}, along with an optional context, and transforms it into a set of PSKs
+and imported identities for use in a connection based on supported (target) protocols and KDFs. In particular,
+for each supported target protocol `target_protocol` and KDF `target_kdf`, the importer constructs
+an ImportedIdentity structure as follows:
 
 ~~~
-   struct {
-       opaque external_identity<1...2^16-1>;
-       opaque label<0..2^8-1>;
-       opaque context<0..2^16-1>;
-       HashAlgorithm hash;
-   } ImportedIdentity;
+uint8 KDFID[2];
+
+struct {
+   opaque external_identity<1...2^16-1>;
+   opaque context<0..2^16-1>;
+   uint16 target_protocol;
+   KDFID target_kdf;
+} ImportedIdentity;
 ~~~
 
-[[TODO: An alternative design might combine label and hash into the same field so that future
-protocols which don't have a notion of HashAlgorithm don't need this field.]]
+The list of KDFID values is maintained by IANA as described in {{IANA}}. External PSKs MUST NOT
+be imported for versions of (D)TLS 1.2 or prior versions. See {{rollout}} for discussion on
+how imported PSKs for TLS 1.3 and non-imported PSKs for earlier versions co-exist for incremental
+deployment.
 
-ImportedIdentity.label MUST be bound to the protocol for which the key is imported. Thus,
-TLS 1.3 and QUICv1 {{!I-D.ietf-quic-transport}} MUST use "tls13" as the label. Similarly, TLS 1.2 and
-all prior TLS versions should use "tls12" as ImportedIdentity.label, as well as SHA256 as ImportedIdentity.hash.
-Note that this means future versions of TLS will increase the number of PSKs derived from an external PSK.
+ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.
+If the EPSK is a key derived from some other protocol or sequence of protocols,
+ImportedIdentity.context MUST include a channel binding for the deriving protocols
+{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included
+in ImportedIdentity.context {{CCB}}.
```

Comment:
This last bit is the kicker for using ImportedIdentity in a handshake.  We should include those secrets in the derivation of the base key, and we might use those secrets in the derivation of the context, but we shouldn't be transmitting them.

Reply:
I'm just removing that sentence (for now).